### DISCUSSION TITLE
Isn't there overhead to create model each time in example?

### DISCUSSION THREAD

**Discussion by wztdream** [google/flax#1559] [Q&A] [answered]
Hi,
Below is the code from Mnist example, you see the `CNN()` module is created with each call of `loss_fn`, is this a good design? it seems to me this will introduce overhead, say we call `loss_fn` 100000 times, we need to create `CNN` that much of times. Or maybe I am miss something here?
```
# flax/examples/mnist/train.py line 56
  def loss_fn(params):
    logits = CNN().apply({'params': params}, images)
    one_hot = jax.nn.one_hot(labels, 10)
    loss = jnp.mean(optax.softmax_cross_entropy(logits=logits, labels=one_hot))
    return loss, logits
```

There is another example related with `BatchNorm` in _ImageNet_ example, you see here the `norm` layer is created in the `__call__` function, so it seems it will create `BatchNorm` each time we call this function, isn't this a waste to create the model each time and through it away each time? 

Note: there is a variable `train` in the `partial` function, so we can not reuse the previously created BatchNorm if the value of `train` changes
```
# flax/examples/imagenet/models.py, line 91
@nn.compact
  def __call__(self, x, train: bool = True):
    conv = partial(nn.Conv, use_bias=False, dtype=self.dtype)
    norm = partial(nn.BatchNorm,
                   use_running_average=not train,
                   momentum=0.9,
                   epsilon=1e-5,
                   dtype=self.dtype)

    x = conv(self.num_filters, (7, 7), (2, 2),
             padding=[(3, 3), (3, 3)],
             name='conv_init')(x)
    x = norm(name='bn_init')(x)
```

**Comment by jheek** [comment, marked as answer]
Due to JAX's compilation with jit/pmap we effectively call loss_fn just once. After that the function is compiled and all Python overhead disappears.

Both JAX and Flax can have large overhead when left uncompiled though. In the future we might try to improve this but typically we find that compilation provides big gains at little costs making eager mode rarely used. My guess is that the construction here wouldn't be a very big factor though. More likely it would be our internal bookkeeping/safety checks that would become the bottleneck

**Reply by wztdream** [reply]
> Due to JAX's compilation with jit/pmap we effectively call loss_fn just once. After that the function is compiled and all Python overhead disappears.
Thank you, it makes senses for the Mnist example. But how about he `BatchNorm` example? How jax deal with the `use_running_average=not train` parameter? Say if `train` changed from `True` to `False`, jax will recompile the BatchNorm model? If `train` keeps `True` or `False` jax will reuse the compiled code, right?

**Reply by jheek** [reply]
Yes typically you will compile twice so for train and eval. This is desired though. The two programs are very different and result in very different compiled programs.

**Reply by wztdream** [reply]
> Yes typically you will compile twice so for train and eval. This is desired though. The two programs are very different and result in very different compiled programs.

Thank you, to make it clear I need ask one more question: we usually need switch between train and eval mode many times, say  **(a)** train 1000 steps-> **(b)** eval 1000 steps(b) --> **(c)**  train 1000 steps --> **(d)** eval 1000 steps and so on. Question is: in step **(c)** does jax reuse previous compiled train-mode model(complied in step (a) I mean) or does jax recompile the model? Same question for **(d)** step for the eval mode? I guess jax will reuse previous compiled train and eval model, so results in only two times of compile,  am I right?

**Reply by jheek** [reply]
>> guess jax will reuse previous compiled train and eval model, so results in only two times of compile, am I right?

That's correct.

**Reply by hhyurdagul** [reply]
> Due to JAX's compilation with jit/pmap we effectively call loss_fn just once. After that the function is compiled and all Python overhead disappears.
> 
> Both JAX and Flax can have large overhead when left uncompiled though. In the future we might try to improve this but typically we find that compilation provides big gains at little costs making eager mode rarely used. My guess is that the construction here wouldn't be a very big factor though. More likely it would be our internal bookkeeping/safety checks that would become the bottleneck

Can you correct me If I am wrong, I understand that the function itself is compiled once, and after that the compiled version runs on consecutive runs, if that is correct is it not a bad thing to run a program(compiled) 10000 times even if it is compiled, does not that slow the process?
